mixer/pkg/attribute/envoyProtoBag.go

```diff
@@ -0,0 +1,316 @@
+	referencedAttrsMutex sync.Mutex
+}
+
+// referencedAttrsSize is the size of referenced attributes.
+
+var envoyProtoBags = sync.Pool{
+	New: func() interface{} {
+		return &EnvoyProtoBag{
+			referencedAttrs: make(map[attr.Reference]attr.Presence, referencedAttrsSize),
+		}
+	},
+}
+
+// GetProtoBag returns a proto-based attribute bag.
+// When you are done using the proto bag, call the Done method to recycle it.
+func GetEnvoyProtoBagAuthz(req *authzGRPC.CheckRequest) *EnvoyProtoBag {
+	//, globalDict map[string]int32,	globalWordList []string) *EnvoyProtoBag {
+	pb := envoyProtoBags.Get().(*EnvoyProtoBag)
+
+	// build the message-level dictionary
+	reqMap := make(map[string]interface{})
+	//TODO account for other protocols
+	reqMap["context.protocol"] = "http"
+	//req_map["context.proxy_version"] = "unknown"
+	reqMap["context.reporter.kind"] = "server"
+	//req_map["context.reporter.uid"] = "unknown"
+	reqMap["destination.ip"] = []byte(net.ParseIP(req.GetAttributes().GetDestination().GetAddress().GetSocketAddress().
+		GetAddress()).To16())
+	reqMap["target.ip"] = []byte(net.ParseIP(req.GetAttributes().GetDestination().GetAddress().GetSocketAddress().
+		GetAddress()).To16())
+	reqMap["destination.port"] = int64(req.GetAttributes().GetDestination().GetAddress().GetSocketAddress().
+		GetPortValue())
+	reqMap["destination.principal"] = req.GetAttributes().GetDestination().GetPrincipal()
+	reqMap["source.principal"] = req.GetAttributes().GetSource().GetPrincipal()
+	reqMap["source.ip"] = []byte(net.ParseIP(req.GetAttributes().GetSource().GetAddress().GetSocketAddress().
+		GetAddress()).To16())
+	reqMap["origin.ip"] = []byte(net.ParseIP(req.GetAttributes().GetSource().GetAddress().GetSocketAddress().
+		GetAddress()).To16())
+	reqMap["source.port"] = int64(req.GetAttributes().GetSource().GetAddress().GetSocketAddress().GetPortValue())
+	//reqMap["source.uid"] = ""
+	//reqMap["destination.uid"] = ""
+	reqMap["request.headers"] = req.GetAttributes().GetRequest().GetHttp().GetHeaders()
+	reqMap["request.host"] = req.GetAttributes().GetRequest().GetHttp().GetHost()
+	reqMap["request.method"] = req.GetAttributes().GetRequest().GetHttp().GetMethod()
+	reqMap["request.path"] = req.GetAttributes().GetRequest().GetHttp().GetPath()
+	reqMap["request.scheme"] = req.GetAttributes().GetRequest().GetHttp().GetScheme()
+	reqMap["request.time"] = time.Unix(int64(req.GetAttributes().GetRequest().GetTime().Seconds),
+		int64(req.GetAttributes().GetRequest().GetTime().Nanos))
+	//reqMap["request.url_path"] = "unknown"
+	reqMap["request.useragent"] = req.GetAttributes().GetRequest().GetHttp().GetHeaders()["user-agent"]
+
+
+	pb.reqMap = reqMap
+
+
+
+	scope.Debugf("Returning bag with attributes:\n%v", pb)
+
+	return pb
+}
+
+// GetProtoBag returns a proto-based attribute bag.
+// When you are done using the proto bag, call the Done method to recycle it.
+func GetEnvoyProtoBagAccessLog(msg *accessLogGRPC.StreamAccessLogsMessage, num int) *EnvoyProtoBag {
```

Comment:
msg.GetHttpLogs().GetLogEntry() returns a list LogEntries. It is because ALS sends a batch in one message. num represents the index of the entry in the batch to get. I can rename it to entryIndex of entryNum if that is more clear.